src/org/wikibase/WikibaseEntityFactory.java

```diff
@@ -0,0 +1,184 @@
+import javax.xml.xpath.XPathFactory;
+
+import org.w3c.dom.DOMException;
+import org.w3c.dom.Document;
+import org.w3c.dom.NamedNodeMap;
+import org.w3c.dom.Node;
+import org.wikibase.data.Entity;
+import org.wikibase.data.Property;
+import org.wikibase.data.Sitelink;
+import org.xml.sax.SAXException;
+
+public class WikibaseEntityFactory {
+    private static Map<String, Entity> items = new HashMap<String, Entity>();
+    private static Pattern ENTITY_ID_PATTERN = Pattern.compile("\\s*(q|Q)?(\\d+)\\s*");
+
+    public static Entity getWikibaseItem(String text) throws WikibaseException {
+        Node entityNode = null;
+        try {
+            DocumentBuilderFactory domBuilderFactory = DocumentBuilderFactory.newInstance();
+            DocumentBuilder builder = domBuilderFactory.newDocumentBuilder();
+            Document document = builder.parse(new ByteArrayInputStream(text.getBytes()));
+
+            XPathFactory xpathFactory = XPathFactory.newInstance();
+            XPath xPath = xpathFactory.newXPath();
+            XPathExpression apiExpression = xPath.compile("/api[1]");
+
+            Node apiNode = (Node) apiExpression.evaluate(document, XPathConstants.NODE);
+
+            if (null == apiNode) {
+                throw new WikibaseException("API root node not found in text.");
+            }
+
+            NamedNodeMap apiAttrs = apiNode.getAttributes();
+            Node successAttr = apiAttrs.getNamedItem("success");
+            String successStatus = successAttr.getTextContent();
+            if (!"1".equals(successStatus)) {
+                throw new WikibaseException("API call unsuccessful.");
+            }
+
+            Node entitiesNode = null;
+            Node eachApiChild = apiNode.getFirstChild();
+            while (null != eachApiChild) {
+                if ("entities".equals(eachApiChild.getNodeName())) {
+                    entitiesNode = eachApiChild;
+                    break;
+                }
+                eachApiChild = eachApiChild.getNextSibling();
+            }
+
+            if (null == entitiesNode) {
+                throw new WikibaseException("Wikibase entities node not found.");
+            }
+            entityNode = entitiesNode.getFirstChild();
+            if (null == entityNode) {
+                throw new WikibaseException("No entity found.");
+            }
+
+            // parse entity attrs
+            NamedNodeMap entityAttrs = entityNode.getAttributes();
+            String itemId = entityAttrs.getNamedItem("title").getNodeValue();
+            Matcher idMatcher = ENTITY_ID_PATTERN.matcher(itemId);
+            if (!idMatcher.matches()) {
+                throw new WikibaseException("String " + itemId + " is not a valid entity ID");
+            }
+            Entity entity = new Entity("P" + idMatcher.group(2));
+            loadItem(entity, entityNode);
+            return entity;
+        } catch (XPathExpressionException e) {
```

Comment:
https://docs.oracle.com/javase/7/docs/technotes/guides/language/catch-multiple.html

Is the IOException resulting from a network error? If so, you should not turn it into a WikibaseException.